Sprint 9 solution http api

Zaza-b1t opened the pull request.

Zaza-b1t committed:
Реализовано ТЗ: менеджер задач с Epic и Subtask

Zaza-b1t committed:
Исправление checkstyle: имя пакета и интервал между методами

Zaza-b1t committed:
Исправление checkstyle

Zaza-b1t committed:
Исправил метод createSubtask() в InMemoryTaskManager

Zaza-b1t committed:
Исправил метод createSubtask() в InMemoryTaskManager

Zaza-b1t committed:
Исправил метод epicCannotAddItselfAsSubtask()

Zaza-b1t committed:
Исправлено приведение типа в InMemoryTaskManagerTest

Zaza-b1t committed:
Реализован FileBackedTaskManager: автосохранение задач в файл

Zaza-b1t committed:
Добавлен приватный конструктор, loadFromFile и ManagerLoadException

Zaza-b1t committed:
Реализовано ТЗ спринта 8: duration, startTime, проверки пересечений

Zaza-b1t committed:
Реализовано ТЗ спринта 8: duration, startTime, проверки пересечений

Zaza-b1t committed:
Исправлены ошибки с CheckStyle

Zaza-b1t committed:
Исправлены ошибки с CheckStyle

Zaza-b1t committed:
Исправлены ошибки с CheckStyle в Task

Zaza-b1t committed:
Исправлены ошибки с CheckStyle в FileBackedTaskManager

Zaza-b1t committed:
Исправлены ошибка в фиксированном име файла FileBackedTaskManagerTest

Zaza-b1t committed:
Исправлены: исключения, интерфейсы, стримы, тесты

Zaza-b1t committed:
Исправлены ошкиби CheckStyle

Zaza-b1t committed:
сделал исключение TaskOverlapException - проверяемым

Zaza-b1t committed:
Выполнил спринт 9

Zaza-b1t committed:
Исправлены ошкиби CheckStyle

Zaza-b1t committed:
Исправлены ошкиби CheckStyle

Zaza-b1t committed:
Исправлены ошкиби CheckStyle

Zaza-b1t committed:
Исправлены ошкиби Run tests

EugeneLenkevich commented on src/server/HistoryHandler.java:
```diff
@@ -0,0 +1,36 @@
+
+public class HistoryHandler extends BaseHttpHandler implements HttpHandler {
+    private final TaskManager taskManager;
+    private final Gson gson = new GsonBuilder()
```
Так и просится вынести создание сериализатора gson в отдельный метод и даже статическое поле какого то спец-класса и использовать везде по коду. Так мы не будем его пересоздавать в каждом классе и, плюс, у нас будет вся настройка сериализации производиться в одном месте. Иначе, поправив настройки только в одном месте, у нас разные части системы могут начать работать по разному :)

EugeneLenkevich commented on test/httpserver/HttpTaskManagerEpicsTest.java:
```diff
@@ -0,0 +1,129 @@
+
+public class HttpTaskManagerEpicsTest {
+    TaskManager manager = new InMemoryTaskManager(Managers.getDefaultHistory());
+    HttpTaskServer taskServer = new HttpTaskServer(manager);
```
А давай всю инициализацию поместим в методы инициализации, `@Before*`, так будет единообразно :)

EugeneLenkevich commented on test/httpserver/HttpTaskManagerEpicsTest.java:
```diff
@@ -0,0 +1,129 @@
+
+        HttpClient client = HttpClient.newHttpClient();
+        URI url = URI.create("http://localhost:8080/epics");
+        HttpRequest request = HttpRequest.newBuilder()
```
Создание и отправка http-запроса по своей сути технические детали, а что если эти технические детали спрятать в методы? Логика тут такая. Тест - это своего рода описание сценария использования системы, именно по ним скорее всего другие разработчики будут разбираться и проверять как работает система, потому и лучше кода это описание легко читается. Потому лучше всю техническую историю спрятать, попробуй, посмотрим что получится :)

EugeneLenkevich commented on test/httpserver/HttpTaskManagerEpicsTest.java:
```diff
@@ -0,0 +1,129 @@
+                .build();
+        HttpResponse<String> getResponse = client.send(getRequest, HttpResponse.BodyHandlers.ofString());
+
+        assertEquals(200, getResponse.statusCode());
```
200

А не хочешь для всех этих http-кодов и http-методов использовать константы, можно даже поискать готовые, например, в подключенных библиотеках. Но руками тоже норм, на твое усмотрение :)

https://www.baeldung.com/java-enum-http-status

EugeneLenkevich reviewed (changes requested):
Привет, Илья! Спасибо за PR.

Отличная работа! :)
Есть совсем небольшие пометки, посмотри комментарии ниже, пожалуйста
В первую очередь хотелось бы 
- посмотреть на читаемость тестов
- унифицировать создание Gson

Zaza-b1t committed:
- Вынес gson в отдельный класс GsonConfig, теперь сериализация везде одинаковая.
- Сделал общий базовый класс BaseHttpTest для тестов, чтобы не дублировать код.
- Добавил enum HttpStatus и константы для URI.
- Тесты теперь проще и чище, техническая логика спрятана в методы.

Zaza-b1t committed:
Исправлены ошкиби Run tests

EugeneLenkevich reviewed (approved):
Во, насколько в итоге проще тест выглядит, супер, теперь все отлично, спасибо :)
